terra/packages/vaa_payload/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+//! should be seen as an explanation of how things are implemented *in this crate* and not as
+//! official documentation.
+//!
+//! ## Wire format
+//!
+//! The VAA payload wire format is not a self-describing format (unlike json and toml). Therefore it
+//! is necessary to know the type that needs to be produced before deserializing a byte stream.
+//!
+//! The wire format currently supports the following primitive types:
+//!
+//! ### `bool`
+//!
+//! Encoded as a single byte where a value of 0 indicates false and 1 indicates true. All other
+//! values are invalid.
+//!
+//! ### Integers
+//!
+//! `i8`, `i16`, `i32`, `i64`, `i128`, `u8`, `u16`, `u32`, `u64`, and `u128` are all supported and
+//! encoded as full-width big-endian integers, i.e., `i16` is 2 bytes, `u64` is 8 bytes, etc.
+//!
+//!
+//! ### `char`
+//!
+//! Encoded as a big-endian `u32`, with the additional restriction that it must be a valid [`Unicode
+//! Scalar Value`](https://www.unicode.org/glossary/#unicode_scalar_value).
+//!
+//! ### Sequences
+//!
+//! Variable length heterogeneous sequences are encoded as a single byte length followed by the
+//! concatenation of the serialized form of each element in the sequence. Note that this means that
+//! sequences cannot have more than 255 elements. Additionally, during serialization the length must
+//! be known ahead of time.
+//!
+//! ### Byte arrays - `&[u8]`, `Vec<u8>`, and `Cow<'a, [u8]>`
+//!
+//! Byte arrays are treated as a subset of variable-length sequences and are encoded as a single
+//! byte length followed by that many bytes of data. Again, since the length of the byte array has
+//! to fit in a single byte it cannot be longer than 255 bytes.
+//!
+//! ### `&str`, `String`
+//!
+//! String types are not currently supported though they could be easily supported by treating them
+//! as byte arrays with a restriction that the byte array must be valid UTF-8.
+//!
+//! ### Tuples
+//!
+//! Tuples are heterogenous sequences where the length is fixed and known ahead of time. In this
+//! case the length is not encoded on the wire and the serialization of each element in the tuple is
+//! concatenated to produce the final value.
+//!
+//! ### `Option<T>`
+//!
+//! The wire format does not support optional values. Options are always deserialized as `Some(T)`
+//! while trying to serialize an `Option::None` will result in an error.
+//!
+//! ### Structs
+//!
+//! Structs are represented the same way as tuples and the wire format for a struct is identical to
+//! the wire format for a tuple with the same fields in the same order. The only exception is unit
+//! structs (structs with no fields), which are not represented in the wire format at all.
+//!
+//! ### `[T; N]`
+//!
+//! Arrays are treated as tuples with homogenous fields and have the same wire format.
+//!
+//! ### Enums
+//!
+//! Enums are encoded as a single byte identifying the variant followed by the serialization of the
```

Comment:
What's the rationale for this encoding? Did you find any VAA types that encode enums in this way? The only enum-like thing that comes to mind is the choice between the different types of payloads, such as "token bridge contract upgrade" and "nft bridge transfer", but discriminating between these is not done by an encoded enum tag, or any other identifier. In fact, there's no explicit discriminator, and I'm not even sure if the VAA grammar used just by the token bridge itself is unambiguous. Even if it's unambiguous, an arbitrary VAA certainly can't be parsed into an appropriate payload type without backtracking.

Reply:
Actually, I take that back. The token bridge and NFT bridge payloads are encoded in this way.